src/Microsoft.ML.Data/Transforms/CopyColumnsTransform.cs

```diff
@@ -47,120 +136,220 @@ public sealed class Arguments : TransformInputBase
+        private readonly Dictionary<int, int> _colNewToOldMapping;
+        private readonly Dictionary<int, int> _colOldToNewMapping;
+        private (string Source, string Name)[] _columns;
+        private readonly IHost _host;
+        public const string LoaderSignature = "CopyColumnsRowMapper";
+
+        private static VersionInfo GetVersionInfo()
+        {
+            return new VersionInfo(
+                modelSignature: "COPYROWM",
+                verWrittenCur: 0x00010001, // Initial
+                verReadableCur: 0x00010001,
+                verWeCanReadBack: 0x00010001,
+                loaderSignature: LoaderSignature);
         }
 
-        public static CopyColumnsTransform Create(IHostEnvironment env, ModelLoadContext ctx, IDataView input)
+        public static CopyColumnsRowMapper Create(IHostEnvironment env, ModelLoadContext ctx, ISchema schema)
         {
             Contracts.CheckValue(env, nameof(env));
             env.CheckValue(ctx, nameof(ctx));
             ctx.CheckAtModel(GetVersionInfo());
-            env.CheckValue(input, nameof(input));
 
             // *** Binary format ***
-            // <handled in ctors>
-            var h = env.Register(RegistrationName);
-            return h.Apply("Loading Model", ch => new CopyColumnsTransform(h, ctx, input));
+            // int: number of added columns
+            // for each added column
+            //   string: output column name
+            //   string: input column name
+
+            var length = ctx.Reader.ReadInt32();
+            var columns = new (string Source, string Name)[length];
+            for (int i = 0; i < length; i++)
+            {
+                columns[i].Name = ctx.LoadNonEmptyString();
+                columns[i].Source = ctx.LoadNonEmptyString();
+            }
+            return new CopyColumnsRowMapper(env, schema, columns);
         }
 
-        public override void Save(ModelSaveContext ctx)
+        public CopyColumnsRowMapper(IHostEnvironment env, ISchema schema, (string Source, string Name)[] columns)
         {
-            Host.CheckValue(ctx, nameof(ctx));
-            ctx.CheckAtModel();
-            ctx.SetVersionInfo(GetVersionInfo());
+            _host = env.Register(LoaderSignature);
+            env.CheckValue(schema, nameof(schema));
+            env.CheckValue(columns, nameof(columns));
+            _schema = schema;
+            _columns = columns;
+            _colNewToOldMapping = new Dictionary<int, int>();
+            _colOldToNewMapping = new Dictionary<int, int>();
+            for (int i = 0; i < _columns.Length; i++)
+            {
+                if (!_schema.TryGetColumnIndex(_columns[i].Source, out int colIndex))
+                {
+                    throw _host.ExceptParam(nameof(schema), $"{_columns[i].Source} not found in {nameof(schema)}");
+                }
+                _colNewToOldMapping.Add(i, colIndex);
+                _colOldToNewMapping.Add(i, colIndex);
```

Comment:
>_colOldToNewMapping.Add(i, colIndex); [](start = 16, length = 37)

I suppose you meant the reverse?

Also note that the same source can map to more than one target. I think it's better to have the new-to-old only (which is one to one), and build a bool array of dependencies inside GetDependencies